File: scripts/create_repo.py
#!/usr/bin/env python3
"""
Create a pristine repository on GitHub or GitLab.

Disables every non-essential feature by default, preserving a pure Git
repository structure unless explicitly enabled via command-line options.
"""

import argparse
from dataclasses import dataclass
import json
import os
import subprocess
import sys
from typing import Any, Dict, Final, List, NoReturn, Optional, Tuple
import urllib.error
import urllib.request

# --- Magic Values Extracted to Constants ---
GITHUB_API_URL: Final[str] = "https://api.github.com/user/repos"
GITHUB_API_VERSION: Final[str] = "2022-11-28"
GITHUB_ACCEPT_HEADER: Final[str] = "application/vnd.github+json"

GITLAB_DEFAULT_BASE_URL: Final[str] = "https://gitlab.com"
GITLAB_PROJECTS_API_PATH: Final[str] = "/api/v4/projects"

HTTP_POST_METHOD: Final[str] = "POST"
TEXT_ENCODING: Final[str] = "utf-8"
CONTENT_TYPE_JSON: Final[str] = "application/json"

GITLAB_FEATURE_ENABLED: Final[str] = "enabled"
GITLAB_FEATURE_DISABLED: Final[str] = "disabled"

STATUS_ENABLED_ICON: Final[str] = "🟩 Enabled"
STATUS_DISABLED_ICON: Final[str] = "🟥 Disabled"


def exit_with_fatal_error(message: str) -> NoReturn:
    print(f"❌ Error: {message}", file=sys.stderr)
    sys.exit(1)


def retrieve_secret_token(token_name: str) -> Optional[str]:
    """Retrieve a credential token from environment variables or secure keyrings."""
    env_token = os.getenv(token_name)
    if env_token:
        return env_token

    if sys.platform == "darwin":
        try:
            darwin_result = subprocess.run(
                ["security", "find-generic-password", "-s", token_name, "-w"],
                capture_output=True,
                text=True,
                check=True,
            )
            return darwin_result.stdout.strip()
        except subprocess.CalledProcessError:
            return None

    if sys.platform.startswith("linux"):
        try:
            linux_result = subprocess.run(
                ["secret-tool", "lookup", "service", token_name],
                capture_output=True,
                text=True,
                check=True,
            )
            return linux_result.stdout.strip()
        except subprocess.CalledProcessError:
            return None

    return None


def execute_http_post_request(
    target_url: str,
    payload: Dict[str, Any],
    request_headers: Dict[str, str],
) -> Dict[str, Any]:
    serialized_data = json.dumps(payload).encode(TEXT_ENCODING)
    http_request = urllib.request.Request(
        target_url,
        data=serialized_data,
        headers=request_headers,
        method=HTTP_POST_METHOD,
    )

    try:
        with urllib.request.urlopen(http_request) as response:
            raw_response_bytes = response.read()
            return json.loads(raw_response_bytes.decode(TEXT_ENCODING))
    except urllib.error.HTTPError as http_error:
        error_body = http_error.read().decode(TEXT_ENCODING)
        exit_with_fatal_error(
            f"API request failed with {http_error.code} {http_error.reason}\n"
            f"Details: {error_body}"
        )
    except urllib.error.URLError as url_error:
        exit_with_fatal_error(f"Network error occurred: {url_error.reason}")


@dataclass(frozen=True)
class GitHubConfiguration:
    repository_name: str
    repository_description: str
    is_public: bool
    enable_issues: bool
    enable_projects: bool
    enable_wiki: bool
    enable_discussions: bool
    enable_downloads: bool


@dataclass(frozen=True)
class GitLabConfiguration:
    project_name: str
    project_description: str
    is_public: bool
    enable_issues: bool
    enable_wiki: bool
    enable_snippets: bool
    enable_merge_requests: bool
    enable_pipelines: bool


def resolve_gitlab_access_level(is_feature_enabled: bool) -> str:
    if is_feature_enabled:
        return GITLAB_FEATURE_ENABLED
    return GITLAB_FEATURE_DISABLED


def create_github_repository(configuration: GitHubConfiguration) -> None:
    api_token = retrieve_secret_token("GITHUB_TOKEN")
    if not api_token:
        exit_with_fatal_error(
            "GITHUB_TOKEN environment variable or Keychain entry is missing."
        )

    # Ensure the Git repository is created without default initial commits
    request_payload: Dict[str, Any] = {
        "name": configuration.repository_name,
        "description": configuration.repository_description,
        "private": not configuration.is_public,
        "auto_init": False,
        "has_issues": configuration.enable_issues,
        "has_projects": configuration.enable_projects,
        "has_wiki": configuration.enable_wiki,
        "has_discussions": configuration.enable_discussions,
        "has_downloads": configuration.enable_downloads,
    }

    request_headers: Dict[str, str] = {
        "Authorization": f"Bearer {api_token}",
        "Accept": GITHUB_ACCEPT_HEADER,
        "X-GitHub-Api-Version": GITHUB_API_VERSION,
        "Content-Type": CONTENT_TYPE_JSON,
    }

    print(f"🚀 Creating pristine GitHub repository '{configuration.repository_name}'...")
    response_data = execute_http_post_request(
        GITHUB_API_URL, request_payload, request_headers
    )

    repository_url = response_data.get("html_url")
    print(f"\n✅ Success! Repository created at: {repository_url}")
    print("\nFeature Status:")

    github_features: List[Tuple[str, str]] = [
        ("has_issues", "Issues"),
        ("has_projects", "Projects"),
        ("has_wiki", "Wiki"),
        ("has_discussions", "Discussions"),
        ("has_downloads", "Downloads"),
    ]

    for api_field, display_name in github_features:
        is_feature_enabled = bool(response_data.get(api_field))
        status_label = (
            STATUS_ENABLED_ICON if is_feature_enabled else STATUS_DISABLED_ICON
        )
        print(f"  {display_name:<12} -> {status_label}")


def create_gitlab_project(configuration: GitLabConfiguration) -> None:
    api_token = retrieve_secret_token("GITLAB_TOKEN")
    if not api_token:
        exit_with_fatal_error(
            "GITLAB_TOKEN environment variable or Keychain entry is missing."
        )

    configured_base_url = os.getenv("GITLAB_URL", GITLAB_DEFAULT_BASE_URL).rstrip("/")
    target_endpoint = f"{configured_base_url}{GITLAB_PROJECTS_API_PATH}"

    # Lock down all built-in GitLab modules by default to maintain pristine state
    request_payload: Dict[str, Any] = {
        "name": configuration.project_name,
        "description": configuration.project_description,
        "visibility": "public" if configuration.is_public else "private",
        "repository_access_level": GITLAB_FEATURE_ENABLED,
        "issues_access_level": resolve_gitlab_access_level(
            configuration.enable_issues
        ),
        "wiki_access_level": resolve_gitlab_access_level(
            configuration.enable_wiki
        ),
        "snippets_access_level": resolve_gitlab_access_level(
            configuration.enable_snippets
        ),
        "merge_requests_access_level": resolve_gitlab_access_level(
            configuration.enable_merge_requests
        ),
        "builds_access_level": resolve_gitlab_access_level(
            configuration.enable_pipelines
        ),
        "analytics_access_level": GITLAB_FEATURE_DISABLED,
        "container_registry_access_level": GITLAB_FEATURE_DISABLED,
        "environments_access_level": GITLAB_FEATURE_DISABLED,
        "feature_flags_access_level": GITLAB_FEATURE_DISABLED,
        "forking_access_level": GITLAB_FEATURE_DISABLED,
        "infrastructure_access_level": GITLAB_FEATURE_DISABLED,
        "model_experiments_access_level": GITLAB_FEATURE_DISABLED,
        "model_registry_access_level": GITLAB_FEATURE_DISABLED,
        "monitor_access_level": GITLAB_FEATURE_DISABLED,
        "package_registry_access_level": GITLAB_FEATURE_DISABLED,
        "pages_access_level": GITLAB_FEATURE_DISABLED,
        "releases_access_level": GITLAB_FEATURE_DISABLED,
        "requirements_access_level": GITLAB_FEATURE_DISABLED,
        "security_and_compliance_access_level": GITLAB_FEATURE_DISABLED,
    }

    request_headers: Dict[str, str] = {
        "Authorization": f"Bearer {api_token}",
        "Content-Type": CONTENT_TYPE_JSON,
    }

    print(f"🚀 Creating pristine GitLab project '{configuration.project_name}'...")
    response_data = execute_http_post_request(
        target_endpoint, request_payload, request_headers
    )

    project_url = response_data.get("web_url")
    print(f"\n✅ Success! Project created at: {project_url}")
    print("\nFeature Status:")

    gitlab_features: List[Tuple[str, str]] = [
        ("Issues", "issues_access_level"),
        ("Wiki", "wiki_access_level"),
        ("Snippets", "snippets_access_level"),
        ("Merge Requests", "merge_requests_access_level"),
        ("Pipelines", "builds_access_level"),
    ]

    for display_name, api_key in gitlab_features:
        is_feature_enabled = response_data.get(api_key) == GITLAB_FEATURE_ENABLED
        status_label = (
            STATUS_ENABLED_ICON if is_feature_enabled else STATUS_DISABLED_ICON
        )
        print(f"  {display_name:<15} -> {status_label}")


def main() -> None:
    main_parser = argparse.ArgumentParser(
        description=(
            "Create a pristine repository on GitHub or GitLab.\n"
            "Disables EVERY feature by default, keeping it a pure Git\n"
            "repository unless explicitly opted-in via command line flags."
        ),
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )

    subparsers = main_parser.add_subparsers(
        dest="platform", required=True, help="Target hosting platform"
    )

    # --- GitHub Subparser ---
    github_parser = subparsers.add_parser(
        "github", help="Create a pure GitHub repository"
    )
    github_parser.add_argument("name", help="Repository name")
    github_parser.add_argument(
        "-d", "--description", default="", help="Repository description"
    )
    github_parser.add_argument(
        "--public",
        action="store_true",
        help="Make repository public (default: private)",
    )

    github_opt_in_group = github_parser.add_argument_group("Opt-in Features")
    github_opt_in_group.add_argument(
        "--issues", action="store_true", help="Enable Issues"
    )
    github_opt_in_group.add_argument(
        "--projects", action="store_true", help="Enable Projects"
    )
    github_opt_in_group.add_argument(
        "--wiki", action="store_true", help="Enable Wiki"
    )
    github_opt_in_group.add_argument(
        "--discussions", action="store_true", help="Enable Discussions"
    )
    github_opt_in_group.add_argument(
        "--downloads", action="store_true", help="Enable Downloads"
    )

    # --- GitLab Subparser ---
    gitlab_parser = subparsers.add_parser(
        "gitlab", help="Create a pure GitLab project"
    )
    gitlab_parser.add_argument("name", help="Project name")
    gitlab_parser.add_argument(
        "-d", "--description", default="", help="Project description"
    )
    gitlab_parser.add_argument(
        "--public",
        action="store_true",
        help="Make project public (default: private)",
    )

    gitlab_opt_in_group = gitlab_parser.add_argument_group("Opt-in Features")
    gitlab_opt_in_group.add_argument(
        "--issues", action="store_true", help="Enable Issues"
    )
    gitlab_opt_in_group.add_argument(
        "--wiki", action="store_true", help="Enable Wiki"
    )
    gitlab_opt_in_group.add_argument(
        "--snippets", action="store_true", help="Enable Snippets"
    )
    gitlab_opt_in_group.add_argument(
        "--merge-requests", action="store_true", help="Enable Merge Requests"
    )
    gitlab_opt_in_group.add_argument(
        "--pipelines", action="store_true", help="Enable CI/CD Pipelines"
    )

    parsed_arguments = main_parser.parse_args()

    match parsed_arguments.platform:
        case "github":
            github_config = GitHubConfiguration(
                repository_name=parsed_arguments.name,
                repository_description=parsed_arguments.description,
                is_public=parsed_arguments.public,
                enable_issues=parsed_arguments.issues,
                enable_projects=parsed_arguments.projects,
                enable_wiki=parsed_arguments.wiki,
                enable_discussions=parsed_arguments.discussions,
                enable_downloads=parsed_arguments.downloads,
            )
            create_github_repository(github_config)
        case "gitlab":
            gitlab_config = GitLabConfiguration(
                project_name=parsed_arguments.name,
                project_description=parsed_arguments.description,
                is_public=parsed_arguments.public,
                enable_issues=parsed_arguments.issues,
                enable_wiki=parsed_arguments.wiki,
                enable_snippets=parsed_arguments.snippets,
                enable_merge_requests=parsed_arguments.merge_requests,
                enable_pipelines=parsed_arguments.pipelines,
            )
            create_gitlab_project(gitlab_config)
        case _:
            exit_with_fatal_error(f"Unsupported platform: {parsed_arguments.platform}")


if __name__ == "__main__":
    main()

File: scripts/create-repo.js
#!/usr/bin/env node

import { execFileSync } from "node:child_process";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

const GITHUB_API_URL = "https://api.github.com/user/repos";
const GITHUB_API_VERSION = "2022-11-28";
const GITHUB_ACCEPT_HEADER = "application/vnd.github+json";

const GITLAB_DEFAULT_BASE_URL = "https://gitlab.com";
const GITLAB_PROJECTS_API_PATH = "/api/v4/projects";

const STATUS_ENABLED_ICON = "🟩 Enabled";
const STATUS_DISABLED_ICON = "🟥 Disabled";

function printHelp() {
  console.log(`
Usage: create-repo <platform> <name> [options]

Platforms:
  github    Create a pristine GitHub repository.
  gitlab    Create a pristine GitLab project.

Arguments:
  name                  Repository / project name.

Options:
  -d, --description <text> Repository description.
      --public          Make repository public (default: private).
      --json            Output API response payload in JSON format.
  -h, --help            Show this help message.

GitHub Opt-in Flags:
      --issues          Enable Issues
      --projects        Enable Projects
      --wiki            Enable Wiki
      --discussions     Enable Discussions
      --downloads       Enable Downloads

GitLab Opt-in Flags:
      --issues          Enable Issues
      --wiki            Enable Wiki
      --snippets        Enable Snippets
      --merge-requests  Enable Merge Requests
      --pipelines       Enable CI/CD Pipelines
      --packages        Enable Package Registry
      --lfs             Enable Git Large File Storage (LFS)

Environment Variables / Keyring Credentials:
  GITHUB_TOKEN          GitHub API Token (or macOS Keychain / Linux secret-tool).
  GITLAB_TOKEN          GitLab API Token (or macOS Keychain / Linux secret-tool).
  GITLAB_URL            GitLab instance base URL (default: https://gitlab.com).

Examples:
  node scripts/create-repo.js github my-repo
  node scripts/create-repo.js github my-app -d "My private repo" --issues
  node scripts/create-repo.js gitlab my-service --public --pipelines
`);
}

function retrieveSecretToken(tokenName) {
  if (process.env[tokenName]) {
    return process.env[tokenName].trim();
  }

  if (process.platform === "darwin") {
    try {
      const output = execFileSync(
        "security",
        ["find-generic-password", "-s", tokenName, "-w"],
        { encoding: "utf8", stdio: ["ignore", "pipe", "ignore"] },
      );
      return output.trim();
    } catch {
      return null;
    }
  }

  if (process.platform === "linux") {
    try {
      const output = execFileSync(
        "secret-tool",
        ["lookup", "service", tokenName],
        {
          encoding: "utf8",
          stdio: ["ignore", "pipe", "ignore"],
        },
      );
      return output.trim();
    } catch {
      return null;
    }
  }

  return null;
}

function resolveGitLabAccessLevel(isEnabled) {
  return isEnabled ? "enabled" : "disabled";
}

function buildGitHubPayload(options) {
  const isPublic = Boolean(options.public);

  return {
    name: options.name,
    description: options.description || "",
    private: !isPublic,
    auto_init: false,
    has_issues: Boolean(options.issues),
    has_projects: Boolean(options.projects),
    has_wiki: Boolean(options.wiki),
    has_discussions: Boolean(options.discussions),
    has_downloads: Boolean(options.downloads),
    is_template: false,
    allow_squash_merge: false,
    allow_rebase_merge: false,
    allow_auto_merge: false,
    delete_branch_on_merge: true,
    allow_forking: false,
    web_commit_signoff_required: true,
    security_and_analysis: {
      advanced_security: { status: "disabled" },
      secret_scanning: { status: "disabled" },
      secret_scanning_push_protection: { status: "disabled" },
      secret_scanning_validity_checks: { status: "disabled" },
      dependabot_security_updates: { status: "disabled" },
    },
  };
}

function buildGitLabPayload(options) {
  const isPublic = Boolean(options.public);
  const repoLevel = isPublic ? "enabled" : "private";

  return {
    name: options.name,
    path: options.name,
    description: options.description || "",
    visibility: isPublic ? "public" : "private",
    repository_access_level: repoLevel,
    issues_access_level: resolveGitLabAccessLevel(options.issues),
    wiki_access_level: resolveGitLabAccessLevel(options.wiki),
    snippets_access_level: resolveGitLabAccessLevel(options.snippets),
    merge_requests_access_level: resolveGitLabAccessLevel(
      options["merge-requests"],
    ),
    builds_access_level: resolveGitLabAccessLevel(options.pipelines),
    forking_access_level: "disabled",
    pages_access_level: "disabled",
    analytics_access_level: "disabled",
    container_registry_access_level: "disabled",
    security_and_compliance_access_level: "disabled",
    releases_access_level: "disabled",
    environments_access_level: "disabled",
    feature_flags_access_level: "disabled",
    infrastructure_access_level: "disabled",
    monitor_access_level: "disabled",
    requirements_access_level: "disabled",
    model_experiments_access_level: "disabled",
    model_registry_access_level: "disabled",
    auto_devops_enabled: false,
    packages_enabled: Boolean(options.packages),
    service_desk_enabled: false,
    lfs_enabled: Boolean(options.lfs),
    shared_runners_enabled: false,
    public_jobs: false,
    emails_disabled: true,
    printing_merge_request_link_enabled: false,
    enforce_auth_checks_on_uploads: true,
    ci_forward_deployment_enabled: false,
    ci_allow_fork_pipelines_to_run_in_parent_project: false,
    container_expiration_policy_attributes: {
      enabled: false,
    },
  };
}

async function executeHttpPost(targetUrl, payload, headers) {
  const response = await fetch(targetUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...headers,
    },
    body: JSON.stringify(payload),
  });

  const responseText = await response.text();
  let responseData;
  try {
    responseData = JSON.parse(responseText);
  } catch {
    responseData = { raw: responseText };
  }

  if (!response.ok) {
    const details = responseData.message
      ? JSON.stringify(responseData.message)
      : responseText;
    throw new Error(
      `API Request Failed (HTTP ${response.status} ${response.statusText}): ${details}`,
    );
  }

  return responseData;
}

async function createGitHubRepo(options) {
  const token = retrieveSecretToken("GITHUB_TOKEN");
  if (!token) {
    throw new Error(
      "GITHUB_TOKEN environment variable or Keychain entry is missing.",
    );
  }

  const payload = buildGitHubPayload(options);
  const headers = {
    Authorization: `Bearer ${token}`,
    Accept: GITHUB_ACCEPT_HEADER,
    "X-GitHub-Api-Version": GITHUB_API_VERSION,
  };

  console.log(`🚀 Creating pristine GitHub repository '${options.name}'...`);
  const response = await executeHttpPost(GITHUB_API_URL, payload, headers);

  if (options.json) {
    console.log(JSON.stringify(response, null, 2));
    return;
  }

  console.log(`\n✅ Success! Repository created at: ${response.html_url}`);
  console.log("\nFeature Status:");

  const features = [
    ["has_issues", "Issues"],
    ["has_projects", "Projects"],
    ["has_wiki", "Wiki"],
    ["has_discussions", "Discussions"],
    ["has_downloads", "Downloads"],
  ];

  for (const [key, label] of features) {
    const isEnabled = Boolean(response[key]);
    const icon = isEnabled ? STATUS_ENABLED_ICON : STATUS_DISABLED_ICON;
    console.log(`  ${label.padEnd(15)} -> ${icon}`);
  }
}

async function createGitLabRepo(options) {
  const token = retrieveSecretToken("GITLAB_TOKEN");
  if (!token) {
    throw new Error(
      "GITLAB_TOKEN environment variable or Keychain entry is missing.",
    );
  }

  const baseUrl = (process.env.GITLAB_URL || GITLAB_DEFAULT_BASE_URL).replace(
    /\/+$/,
    "",
  );
  const targetUrl = `${baseUrl}${GITLAB_PROJECTS_API_PATH}`;

  const payload = buildGitLabPayload(options);
  const headers = {
    Authorization: `Bearer ${token}`,
  };

  console.log(`🚀 Creating pristine GitLab project '${options.name}'...`);
  const response = await executeHttpPost(targetUrl, payload, headers);

  if (options.json) {
    console.log(JSON.stringify(response, null, 2));
    return;
  }

  console.log(`\n✅ Success! Project created at: ${response.web_url}`);
  console.log("\nFeature Status:");

  const features = [
    ["Issues", "issues_access_level"],
    ["Wiki", "wiki_access_level"],
    ["Snippets", "snippets_access_level"],
    ["Merge Requests", "merge_requests_access_level"],
    ["Pipelines", "builds_access_level"],
    ["Packages", "packages_enabled"],
    ["LFS", "lfs_enabled"],
  ];

  for (const [label, key] of features) {
    const isEnabled = response[key] === "enabled" || response[key] === true;
    const icon = isEnabled ? STATUS_ENABLED_ICON : STATUS_DISABLED_ICON;
    console.log(`  ${label.padEnd(15)} -> ${icon}`);
  }
}

async function main() {
  let args;
  try {
    args = parseArgs({
      options: {
        description: { type: "string", short: "d" },
        public: { type: "boolean", default: false },
        issues: { type: "boolean", default: false },
        projects: { type: "boolean", default: false },
        wiki: { type: "boolean", default: false },
        discussions: { type: "boolean", default: false },
        downloads: { type: "boolean", default: false },
        snippets: { type: "boolean", default: false },
        "merge-requests": { type: "boolean", default: false },
        pipelines: { type: "boolean", default: false },
        packages: { type: "boolean", default: false },
        lfs: { type: "boolean", default: false },
        json: { type: "boolean", default: false },
        help: { type: "boolean", short: "h", default: false },
      },
      allowPositionals: true,
    });
  } catch (err) {
    console.error(`CLI Error: ${err.message}`);
    console.error("Run with --help for usage details.");
    process.exit(1);
  }

  const { values, positionals } = args;

  if (values.help) {
    printHelp();
    process.exit(0);
  }

  const platform = positionals[0]?.toLowerCase();
  const repoName = positionals[1];

  if (!platform || !["github", "gitlab"].includes(platform)) {
    console.error(
      "Error: Platform argument ('github' or 'gitlab') is required.",
    );
    printHelp();
    process.exit(1);
  }

  if (!repoName) {
    console.error("Error: Repository / project name argument is required.");
    printHelp();
    process.exit(1);
  }

  const options = {
    ...values,
    name: repoName,
  };

  try {
    if (platform === "github") {
      await createGitHubRepo(options);
    } else {
      await createGitLabRepo(options);
    }
  } catch (err) {
    console.error(`\n❌ Error: ${err.message}`);
    process.exit(1);
  }
}

export {
  buildGitHubPayload,
  buildGitLabPayload,
  resolveGitLabAccessLevel,
  retrieveSecretToken,
};

if (
  process.argv[1] &&
  path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)
) {
  main();
}
